pymecht/MatModel.py

```diff
@@ -1044,4 +1044,136 @@ def _energy_stress(self,A,B,mean_theta,sigma,aniso_fraction,**extra_args):
+    Will be developed based on symbolic differentiation.
+    Psi = <USER-INPUT FUNCTION AS STRING>
+    '''
+    def __init__(self, _W='', _init_guess='', _low_bound='', _up_bound='', dparams=False):
+        super().__init__()
+        # Added success flag to handle if wrong format is given.
+        success = False
+        while success == False:
+            try:
+                # Get initial guess. Needed since we don't know formor number
+                # of parameters.
+                if _init_guess!='':
+                    self.param_default = eval('dict('+_init_guess+')')
+                else:
+                    self.param_default  = eval('dict('+input("Please enter inital guess for parameters in the form PARAM1=VAL1, PARAM2=VAL2, ..., PARAMN=VALN: ")+')')#dict(c1=1.,c2=1.,c3=1.,c4=0.)
+                success = True
+            except:
+                print("Invalid format. Please try again.")
+        success = False
+        while success == False:
+            try:
+                if _low_bound!='':
+                    self.param_low_bd = eval('dict('+_low_bound+')')
+                else:
+                    self.param_low_bd   = eval('dict('+input("Please enter lower bound for parameters in the form PARAM1=VAL1, PARAM2=VAL2, ..., PARAMN=VALN: ")+')')#dict(c1=0.0001,c2=0.,c3=0.,c4=0.)
+                if len(self.param_low_bd) == 0:
+                    print("Warning: Lower bound not set for parameters")
+                success = True
+            except:
+                print("Invalid format. Please try again.")
+        success = False
+        while success == False:
+            try:
+                if _up_bound!='':
+                    self.param_up_bd = eval('dict('+_up_bound+')')
+                else:
+                    self.param_up_bd    = eval('dict('+input("Please enter upper bound for parameters in the form PARAM1=VAL1, PARAM2=VAL2, ..., PARAMN=VALN: ")+')')#dict(c1=100.,c2=100.,c3=100.,c4=100.)
+                if len(self.param_up_bd) == 0:
+                    print("Warning: Upper bound not set for parameters")
+                success = True
+            except:
+                print("Invalid format. Please try again.")
+        
+        # Get the user's strain energy density function (SEDF)
+        if _W!='':
+            self.energy_form = _W
+        else:
+            self.energy_form = input("Please enter form of the SEDF: ")#"c1*(I1-3)+c2*(I1-3)**2+c3*(I1-3)**3+c4*(I1-3)**4"
+        # Since there are an arbitrary number of arbitrarily-named parameters,
+        # we get these from the definition of the initial guess.
+        self.param_names = [i for i in self.param_default]
+        
+        # Need sympy for symbolic differentiation
+        # Derivatives calculated here to avoid runtime being called in loop
+        import sympy as sp
+        
+        # Define invariants as symbols
+        I1,I2,I3,I4 = sp.symbols('I1 I2 I3 I4')
```

Comment:
The third invariant should be J, since that is the third derivative expected to be dPsi/dJ.

Reply:
I've extended this to take both. Better solution or do you prefer it to only be J?